pkg/capabilities/triggers/mercury_trigger.go

```diff
@@ -225,19 +171,85 @@ func (o *MercuryTriggerService) GetTriggerID(req capabilities.CapabilityRequest,
 	return "", fmt.Errorf("triggerId not found in inputs")
 }
 
-func GenerateTriggerEventID(reports []mercury.FeedReport) string {
-	// Let's hash all the feedIds and timestamps together
-	sort.Slice(reports, func(i, j int) bool {
-		if reports[i].FeedID == reports[j].FeedID {
-			return reports[i].ObservationTimestamp < reports[j].ObservationTimestamp
+func (o *MercuryTriggerService) loop() {
+	defer o.wg.Done()
+	now := time.Now().UnixMilli()
+	nextWait := o.tickerResolutionMs - now%o.tickerResolutionMs
+
+	for {
+		select {
+		case <-o.stopCh:
+			return
+		case <-time.After(time.Duration(nextWait) * time.Millisecond):
+			startTs := time.Now().UnixMilli()
+			// find closest timestamp that is a multiple of o.tickerResolutionMs
+			aligned := (startTs + o.tickerResolutionMs/2) / o.tickerResolutionMs * o.tickerResolutionMs
+			o.process(aligned)
+			endTs := time.Now().UnixMilli()
+			if endTs-startTs > o.tickerResolutionMs {
+				o.lggr.Errorw("processing took longer than ticker resolution", "duration", endTs-startTs, "tickerResolutionMs", o.tickerResolutionMs)
+			}
+			nextWait = getNextWaitIntervalMs(aligned, o.tickerResolutionMs, endTs)
+		}
+	}
+}
+
+func getNextWaitIntervalMs(lastTs, tickerResolutionMs, currentTs int64) int64 {
+	desiredNext := lastTs + tickerResolutionMs
+	nextWait := desiredNext - currentTs
+	if nextWait <= 0 {
+		nextWait = 0
+	}
+	return nextWait
+}
+
+func (o *MercuryTriggerService) process(timestamp int64) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	for _, sub := range o.subscribers {
+		if timestamp%int64(sub.config.MaxFrequencyMs) == 0 {
+			reportList := make([]mercury.FeedReport, 0)
+			for _, feedID := range sub.config.FeedIds {
+				if latest, ok := o.latestReports[mercury.FeedID(feedID)]; ok {
+					reportList = append(reportList, latest)
+				}
+			}
+
+			val, err := mercury.Codec{}.Wrap(reportList)
+			if err != nil {
+				o.lggr.Errorw("Error encoding reportList", "err", err)
+				continue
+			}
+
+			timestampStr := strconv.FormatInt(timestamp, 10)
+			// use 32-byte-padded timestamp as EventID (human-readable)
+			paddedTs := fmt.Sprintf("mercury_%024s", timestampStr)
+			triggerEvent := capabilities.TriggerEvent{
+				TriggerType: "mercury",
+				ID:          paddedTs,
+				Timestamp:   timestampStr,
+				Payload:     val,
+			}
+
+			eventVal, err := values.Wrap(triggerEvent)
+			if err != nil {
+				o.lggr.Errorw("Error wrapping triggerEvent", "err", err)
+				continue
+			}
+
+			// Create a new CapabilityResponse with the MercuryTriggerEvent
+			capabilityResponse := capabilities.CapabilityResponse{
+				Value: eventVal,
+			}
+
+			o.lggr.Debugw("ProcessReport pushing event", "nReports", len(reportList), "eventID", triggerEvent.ID)
+			select {
```

Comment:
Hmmmm... Note that the workflow engine currently uses an unbuffered channel, and that this channel is read by a synchronising thread that also does other things. I'm not sure we should make the assumption that the channel will be buffered. It feels like this begs the possibility of message loss, which will lead to consensus issues right?

Reply:
OK then would you rather add buffering on the trigger side? Why is that better than buffering on the engine side?

Reply:
Agreed to re-think and address separately (either in the trigger or in engine - tbd).